drivers/input/inport/pnp.c

```diff
@@ -0,0 +1,283 @@
+                }
+
+                break;
+            }
+
+            default:
+                DPRINT("[%p:%X:%X] Unrecognized resource type %X\n",
+                       Descriptor,
+                       Descriptor->ShareDisposition,
+                       Descriptor->Flags,
+                       Descriptor->Type);
+                break;
+        }
+    }
+
+    if (!FoundBasePort || !FoundIrq)
+    {
+        DPRINT1("The device resources were not found\n");
+        Status = STATUS_DEVICE_CONFIGURATION_ERROR;
+        goto Complete;
+    }
+
+    DPRINT("I/O base at %p\n", DeviceExtension->IoBase);
+    DPRINT("IRQ %d\n", RawVector);
+
+    Status = InPortWmiRegistration(DeviceExtension);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("WMI registration failed 0x%X\n", Status);
+        goto Complete;
+    }
+
+    InPortInitializeMouse(DeviceExtension);
+
+    Status = IoConnectInterrupt(&DeviceExtension->InterruptObject,
+                                InPortIsr,
+                                DeviceExtension,
+                                NULL,
+                                DeviceExtension->InterruptVector,
+                                DeviceExtension->InterruptLevel,
+                                DeviceExtension->InterruptLevel,
+                                DeviceExtension->InterruptMode,
+                                DeviceExtension->InterruptShared,
+                                DeviceExtension->InterruptAffinity,
+                                FALSE);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Could not connect to interrupt %d\n", DeviceExtension->InterruptVector);
+        goto Complete;
+    }
+
+    KeSynchronizeExecution(DeviceExtension->InterruptObject,
+                           InPortStartMouse,
+                           DeviceExtension);
+
+    DeviceExtension->State = dsStarted;
+
+Complete:
+    Irp->IoStatus.Status = Status;
+    IoCompleteRequest(Irp, IO_NO_INCREMENT);
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+InPortRemoveDevice(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _Inout_ PIRP Irp)
+{
+    NTSTATUS Status;
+    BOOLEAN IsStarted;
+    PINPORT_DEVICE_EXTENSION DeviceExtension = DeviceObject->DeviceExtension;
+
+    PAGED_CODE();
+
+    InPortWmiDeRegistration(DeviceExtension);
+
+    IsStarted = (DeviceExtension->State == dsStarted);
+
+    DeviceExtension->State = dsRemoved;
+
+    Irp->IoStatus.Status = STATUS_SUCCESS;
+    IoSkipCurrentIrpStackLocation(Irp);
+    Status = IoCallDriver(DeviceExtension->Ldo, Irp);
+
+    IoReleaseRemoveLockAndWait(&DeviceExtension->RemoveLock, Irp);
+
+    /* Device is active */
+    if (IsStarted)
+    {
+        KeSynchronizeExecution(DeviceExtension->InterruptObject,
+                               InPortStopMouse,
+                               DeviceExtension);
+
+        /* Flush DPC for ISR */
+        KeFlushQueuedDpcs();
+
+        IoDisconnectInterrupt(DeviceExtension->InterruptObject);
```

Comment:
I think this should come before `KeFlushQueuedDpcs`. First we want to guarantee that the ISR is not running (since it could queue a new DPC); _then_ we want to make sure no DPCs are running.